pkg/flow/internal/controller/node_component.go

```diff
@@ -254,6 +254,7 @@ func (cn *ComponentNode) Evaluate(scope *vm.Scope) error {
 	case nil:
 		cn.setEvalHealth(component.HealthTypeHealthy, "component evaluated")
 	default:
+		cn.moduleController.ClearModuleIDs()
```

Comment:
There's a bug here. 

1. A module loader is constructed successfully. 
2. The user updates the module loader with an invalid configuration (i.e., something that will fail to evaluate)
3. Loaded modules are incorrectly cleared. 

I think you only want to do this if the _first_ evaluation fails.

Reply:
I am thinking this might depend on the behavior of the module loader. IE is it an error with the config passed in for example an invalid file path, or is it an error with the content. Which leads me to think going back to the original commit of having the creator of the module know when to remove is the safest path even if it requires more vigilance.

Reply:
The original issue was specifically that modules leak when constructing a component fails, and how that leaks to the next instance of the component. 

I'm personally less worried about leaking modules when a subsequent evaluation fails, especially with the change you've made to Run, since it guarantees that a module registration will never leak into the next instance of the component. 

If a module leaks within the same component between calls to Update, I'd say that's a component bug which should be fixed at the module level; but I think the bug you're fixing here is a controller bug.

Reply:
> I think the bug you're fixing here is a controller bug.

To rephrase this: I think it's wrong for the Flow controller to retain state across two instances of the same component, which you're now fixing.